From 12db386729a50f353f21c6b235fa4a73b411f190 Mon Sep 17 00:00:00 2001
From: LordYuuma <lordyuuma@gmail.com>
Date: Sat, 12 Jun 2021 01:19:44 +0200
Subject: [PATCH] edopro: Respect YGOPRO environment paths.

* premake5.lua (environment-paths): New option.
* gframe/game.cpp (PopulateResourcesDirectories)[YGOPRO_ENVIRONMENT_PATHS]:
Load scripts from YGOPRO_SCRIPT_PATH.
Load pics from YGOPRO_IMAGE_PATH.  Fall back to local downloader.
* gframe/data_handler.cpp (LoadDatabases)[YGOPRO_ENVIRONMENT_PATHS]:
Load databases and LFLists from YGOPRO_DATA_PATH.
* gframe/deck_manager.cpp (LoadLFList)[YGOPRO_ENVIRONMENT_PATHS]:
Don't load local LFLists -- use those loaded by the DataHandler instead.
* gframe/premake5.lua (ygopro-config)[environment-paths]:
Define YGOPRO_ENVIRONMENT_PATHS.
* gframe/utils.h (NormalizePath): Allow absolute paths.
(PathForeach): New function.
* gframe/utils.cpp (PathForeach): Implement it.
* gframe/text_types.h (operator/): Override for path_string.
---
 gframe/data_handler.cpp | 16 ++++++++++++++++
 gframe/deck_manager.cpp |  2 ++
 gframe/game.cpp         | 26 ++++++++++++++++++++++++++
 gframe/premake5.lua     |  4 ++++
 gframe/text_types.h     | 16 +++++++++++++++-
 gframe/utils.cpp        | 11 ++++++++++-
 gframe/utils.h          |  3 ++-
 premake5.lua            |  4 ++++
 8 files changed, 79 insertions(+), 3 deletions(-)

diff --git a/gframe/data_handler.cpp b/gframe/data_handler.cpp
index 3adbc52..56a2f6c 100644
--- a/gframe/data_handler.cpp
+++ b/gframe/data_handler.cpp
@@ -17,10 +17,25 @@
 #include "Android/COSAndroidOperator.h"
 #include "Android/porting_android.h"
 #endif
+#include "deck_manager.h"
 
 namespace ygo {
 
 void DataHandler::LoadDatabases() {
+#ifdef YGOPRO_ENVIRONMENT_PATHS
+	const char* data_path_env = getenv("YGOPRO_DATA_PATH");
+	if(!data_path_env) return;
+	epro::path_string data_path_s = Utils::ToPathString(data_path_env);
+	ygo::Utils::PathForeach(
+		data_path_s,
+		[&](const epro::path_string& prefix) {
+			for (auto& file : Utils::FindFiles(prefix, { EPRO_TEXT("cdb") }))
+				if(dataManager->LoadDB(prefix / file))
+					WindBot::AddDatabase(prefix / file);
+			deckManager->LoadLFListSingle(prefix / EPRO_TEXT("lflist.conf"));
+			dataManager->LoadStrings(prefix / EPRO_TEXT("strings.conf"));
+		});
+#else
 	if(Utils::FileExists(EPRO_TEXT("cards.cdb")) && std::ifstream(EPRO_TEXT("cards.cdb")).good()) {
 		if(dataManager->LoadDB(EPRO_TEXT("cards.cdb")))
 			WindBot::AddDatabase(EPRO_TEXT("cards.cdb"));
@@ -31,6 +46,7 @@ void DataHandler::LoadDatabases() {
 			WindBot::AddDatabase(db);
 	}
 	LoadArchivesDB();
+#endif
 }
 void DataHandler::LoadArchivesDB() {
 	for(auto& archive : Utils::archives) {
diff --git a/gframe/deck_manager.cpp b/gframe/deck_manager.cpp
index 909f5dd..b3edeb5 100644
--- a/gframe/deck_manager.cpp
+++ b/gframe/deck_manager.cpp
@@ -99,9 +99,11 @@ bool DeckManager::LoadLFListFolder(epro::path_string path) {
 	return loaded;
 }
 void DeckManager::LoadLFList() {
+#ifndef YGOPRO_ENVIRONMENT_PATHS
 	LoadLFListSingle(EPRO_TEXT("./expansions/lflist.conf"));
 	LoadLFListSingle(EPRO_TEXT("./lflist.conf"));
 	LoadLFListFolder(EPRO_TEXT("./lflists/"));
+#endif
 	LFList nolimit;
 	nolimit.listName = L"N/A"; // N/A
 	nolimit.hash = 0;
diff --git a/gframe/game.cpp b/gframe/game.cpp
index bf7f440..87233ad 100644
--- a/gframe/game.cpp
+++ b/gframe/game.cpp
@@ -3347,6 +3347,31 @@ void Game::UpdateUnzipBar(unzip_payload* payload) {
 	game->updateProgressBottom->setProgress(payload->percentage);
 }
 void Game::PopulateResourcesDirectories() {
+#ifdef YGOPRO_ENVIRONMENT_PATHS
+	const char* script_path_env = getenv("YGOPRO_SCRIPT_PATH");
+	if (script_path_env)
+		Utils::PathForeach(
+			Utils::ToPathString(script_path_env),
+			[this](const epro::path_string& prefix) {
+				epro::path_string script_dir = Utils::NormalizePath(prefix);
+				script_dirs.push_back(script_dir);
+				auto script_subdirs = Utils::FindSubfolders(script_dir);
+				script_dirs.insert(script_dirs.end(), script_subdirs.begin(), script_subdirs.end());
+			});
+	const char* image_path_env = getenv("YGOPRO_IMAGE_PATH");
+	if(image_path_env)
+		Utils::PathForeach(
+			Utils::ToPathString(image_path_env),
+			[this](const epro::path_string& prefix) {
+				epro::path_string image_dir = Utils::NormalizePath(prefix);
+				pic_dirs.push_back(image_dir);
+				cover_dirs.push_back(image_dir / EPRO_TEXT("cover/"));
+				field_dirs.push_back(image_dir / EPRO_TEXT("field/"));
+			});
+	pic_dirs.push_back(EPRO_TEXT("./pics/"));
+	cover_dirs.push_back(EPRO_TEXT("./pics/cover/"));
+	field_dirs.push_back(EPRO_TEXT("./pics/field/"));
+#else
 	script_dirs.push_back(EPRO_TEXT("./expansions/script/"));
 	auto expansions_subdirs = Utils::FindSubfolders(EPRO_TEXT("./expansions/script/"));
 	script_dirs.insert(script_dirs.end(), std::make_move_iterator(expansions_subdirs.begin()), std::make_move_iterator(expansions_subdirs.end()));
@@ -3363,6 +3388,7 @@ void Game::PopulateResourcesDirectories() {
 	field_dirs.push_back(EPRO_TEXT("./expansions/pics/field/"));
 	field_dirs.push_back(EPRO_TEXT("archives"));
 	field_dirs.push_back(EPRO_TEXT("./pics/field/"));
+#endif
 }
 
 void Game::PopulateLocales() {
diff --git a/gframe/premake5.lua b/gframe/premake5.lua
index 5297693..1a14fb7 100644
--- a/gframe/premake5.lua
+++ b/gframe/premake5.lua
@@ -27,6 +27,10 @@ local ygopro_config=function(static_core)
 	if _OPTIONS["update-url"] then
 		defines { "UPDATE_URL=" .. _OPTIONS["update-url"] }
 	end
+	if _OPTIONS["environment-paths"] then
+		defines { "YGOPRO_ENVIRONMENT_PATHS" }
+	end
+
 	includedirs "../ocgcore"
 	links { "clzma", "freetype", "Irrlicht" }
 	filter "system:macosx"
diff --git a/gframe/text_types.h b/gframe/text_types.h
index c1072c0..d3ac1e3 100644
--- a/gframe/text_types.h
+++ b/gframe/text_types.h
@@ -44,4 +44,18 @@ template<typename T>
 using basic_string_view = nonstd::basic_string_view<T>;
 }
 using namespace nonstd::literals;
-#endif /* TEXT_TYPES_H_ */
\ No newline at end of file
+inline epro::path_string operator/(const epro::path_string& base, epro::path_stringview subdir) {
+	if (base.empty() || base == EPRO_TEXT(".")) {
+		epro::path_string path(subdir);
+		return path;
+	}
+	else {
+		epro::path_string path(base);
+		if (base.back() != EPRO_TEXT('/'))
+			path += EPRO_TEXT("/");
+		path += epro::path_string(subdir);
+		return path;
+	}
+}
+
+#endif /* TEXT_TYPES_H_ */
diff --git a/gframe/utils.cpp b/gframe/utils.cpp
index 5483546..448178f 100644
--- a/gframe/utils.cpp
+++ b/gframe/utils.cpp
@@ -341,6 +341,16 @@ namespace ygo {
 		return true;
 	}
 
+	void Utils::PathForeach(epro::path_stringview path, const std::function<void(epro::path_string)>& cb)
+	{
+		// FIXME: should be ';' in WIN32 and WIN64.
+		static const epro::path_char path_sep = EPRO_TEXT(':');
+		std::basic_istringstream<epro::path_char> dirs(path.data());
+		epro::path_string dir;
+		while (std::getline(dirs, dir, path_sep))
+			cb(dir);
+	}
+
 	epro::path_stringview Utils::GetExePath() {
 		static epro::path_string binarypath = []()->epro::path_string {
 #ifdef _WIN32
@@ -484,4 +494,3 @@ namespace ygo {
 #endif
 	}
 }
-
diff --git a/gframe/utils.h b/gframe/utils.h
index 1fdf817..922a816 100644
--- a/gframe/utils.h
+++ b/gframe/utils.h
@@ -49,7 +49,7 @@ namespace ygo {
 			static_assert(N <= 16, "Thread name on posix can't be more than 16 bytes!");
 			InternalSetThreadName(s, ws);
 		}
-		
+
 		static std::vector<SynchronizedIrrArchive> archives;
 		static irr::io::IFileSystem* filesystem;
 		static irr::IOSOperator* OSOperator;
@@ -69,6 +69,7 @@ namespace ygo {
 		static void CreateResourceFolders();
 		static void FindFiles(epro::path_stringview path, const std::function<void(epro::path_stringview, bool)>& cb);
 		static std::vector<epro::path_string> FindFiles(epro::path_stringview path, const std::vector<epro::path_stringview>& extensions, int subdirectorylayers = 0);
+		static void PathForeach(epro::path_stringview path, const std::function<void(epro::path_string)>& cb);
 		/** Returned subfolder names are prefixed by the provided path */
 		static std::vector<epro::path_string> FindSubfolders(epro::path_stringview path, int subdirectorylayers = 1, bool addparentpath = true);
 		static std::vector<int> FindFiles(irr::io::IFileArchive* archive, epro::path_stringview path, const std::vector<epro::path_stringview>& extensions, int subdirectorylayers = 0);
diff --git a/premake5.lua b/premake5.lua
index 1e71da1..eea655c 100644
--- a/premake5.lua
+++ b/premake5.lua
@@ -60,6 +60,10 @@ newoption {
 	value = "url",
 	description = "API endpoint to check for updates from"
 }
+newoption {
+	trigger = "environment-paths",
+	description = "Read databases, scripts and images from YGOPRO_*_PATH"
+}
 workspace "ygo"
 	location "build"
 	language "C++"
--
libgit2 1.1.0

